plugins/cros_ec-usb/fu-cros_ec-usb-device.c

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright (C) 2020 Benson Leung <bleung@chromium.org>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include "fu-cros_ec-usb-device.h"
+#include "fu-cros_ec-usb-common.h"
+
+
+#define USB_SUBCLASS_GOOGLE_UPDATE 0x53
+#define USB_PROTOCOL_GOOGLE_UPDATE 0xff
+
+struct _FuCrOSECUsbDevice {
+	FuUsbDevice			parent_instance;
+	guint8				iface_idx; 	/* bInterfaceNumber */
+	guint8				ep_num;		/* bEndpointAddress */
+	guint16				chunk_len; 	/* wMaxPacketSize */
+
+	struct first_response_pdu	targ;
+	guint16				protocol_version;
+	guint16				header_type;
+	struct cros_ec_version		version;
+};
+
+
+G_DEFINE_TYPE (FuCrOSECUsbDevice, fu_cros_ec_usb_device, FU_TYPE_USB_DEVICE)
+
+static gboolean
+fu_cros_ec_usb_device_find_interface (FuUsbDevice *device,
+				      GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (device);
+	FuCrOSECUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+	g_autoptr(GPtrArray) intfs = NULL;
+
+	/* based on usb_updater2's find_interfacei() and find_endpoint() */
+
+	intfs = g_usb_device_get_interfaces (usb_device, error);
+	if (intfs == NULL)
+		return FALSE;
+	for (guint i = 0; i < intfs->len; i++) {
+		GUsbInterface *intf = g_ptr_array_index (intfs, i);
+		if (g_usb_interface_get_class (intf) == 255 &&
+		    g_usb_interface_get_subclass (intf) == USB_SUBCLASS_GOOGLE_UPDATE &&
+		    g_usb_interface_get_protocol (intf) == USB_PROTOCOL_GOOGLE_UPDATE) {
+			GUsbEndpoint *ep;
+			g_autoptr(GPtrArray) endpoints = NULL;
+
+			endpoints = g_usb_interface_get_endpoints (intf);
+			if (NULL == endpoints || 0 == endpoints->len)
+				continue;
+			ep = g_ptr_array_index (endpoints, 0);
+			self->iface_idx = g_usb_interface_get_number (intf);
+			self->ep_num = g_usb_endpoint_get_address (ep) & 0x7f;
+			self->chunk_len = g_usb_endpoint_get_maximum_packet_size (ep);
+
+			return TRUE;
+		}
+	}
+	g_set_error_literal (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_FOUND,
+			     "No Google Update interface found.");
+	return FALSE;
+}
+
+static gboolean
+fu_cros_ec_usb_device_open (FuUsbDevice *device, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (device);
+	FuCrOSECUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+
+	if (!g_usb_device_claim_interface (usb_device, self->iface_idx,
+					   G_USB_DEVICE_CLAIM_INTERFACE_BIND_KERNEL_DRIVER,
+					   error)) {
+		g_prefix_error (error, "failed to claim interface: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+
+static gboolean
+fu_cros_ec_usb_device_probe (FuUsbDevice *device, GError **error)
+{
+	FuCrOSECUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+
+	/* very much like usb_updater2's usb_findit() */
+
+	if (!fu_cros_ec_usb_device_find_interface (device, error)) {
+		g_prefix_error (error, "failed to find update interface: ");
+		return FALSE;
+	}
+
+	if (0 == self->chunk_len) {
```

Comment:
by convention, and for no other reason, we tend to do `self->chunk_len == 0`. Not a hill I'm going to die on tho.